### ISSUE TITLE
instanceof / structural typing issue

### ISSUE THREAD

**Issue by mikelehen** [Microsoft/TypeScript#15585]
This may be a dupe of #11664 though that deals with a type union and this is purely inheritance.

**TypeScript Version:**  2.3.2

**Code**
```ts
abstract class Value {
  readonly valueType: number;
  abstract value(): any;
}

class NullValue extends Value {
  valueType = 0;

  value(): any {
    return null;
  }
}

class NumberValue extends Value {
  valueType = 1;

  constructor(readonly num: number) { super(); }

  value(): any {
    return this.num;
  }
}

function process(node: Value) {
  if (node instanceof NullValue) {
    console.log('Null!')
  } else if (node instanceof NumberValue) {
    // Produces: TS2339: Property 'num' does not exist on type 'never'.
    console.log('NumberValue', node.num);
  }
}
```

**Expected behavior:** No errors.

**Actual behavior:** Error:
```
test.ts(29,37): error TS2339: Property 'num' does not exist on type 'never'.
```

I can "solve" the issue by adding a dummy `readonly foo = null;` field to the NullValue class.

**Comment by aluanhaddad** [comment]
The issue here is that, as you suggest, `NullValue` is a structural supertype of `NumberValue`. Since you test for `NullValue` first, the typechecker concludes that that `node` cannot be a `NumberValue` since it did not meet the requirements to be a value of its supertype.

By the same token, the following assignment is legal
```ts
const x : NullValue = new NumberValue(1);
```

However, in the case of the process function above, if one just reverses the order of the checks like so
```ts
function process(node: Value) {
  if (node instanceof NumberValue) {
    console.log('Null!')
  } else if (node instanceof NullValue) {
    console.log('NumberValue', node.num);
  }
}
```
there is no longer an error.
